pkg/cover/backend/dwarf.go

```diff
@@ -363,41 +392,97 @@ func readTextRanges(debugInfo *dwarf.Data, module *vminfo.KernelModule, pcFix pc
 		if ent.Tag != dwarf.TagCompileUnit {
 			return nil, nil, fmt.Errorf("found unexpected tag %v on top level", ent.Tag)
 		}
-		attrName := ent.Val(dwarf.AttrName)
-		if attrName == nil {
+		attrName, ok := ent.Val(dwarf.AttrName).(string)
+		if !ok {
 			continue
 		}
-		unit := &CompileUnit{
-			ObjectUnit: ObjectUnit{
-				Name: attrName.(string),
-			},
-			Module: module,
-		}
-		units = append(units, unit)
-		ranges1, err := debugInfo.Ranges(ent)
-		if err != nil {
-			return nil, nil, err
-		}
+		attrCompDir, _ := ent.Val(dwarf.AttrCompDir).(string)
+		// Compile unit names are relative to the compilation dir, while per-line info isn't.
+		// Let's stick to the common approach.
+		unitName := attrCompDir + attrName
 
-		var filtered bool
-		for _, r := range ranges1 {
-			if pcFix != nil {
-				r, filtered = pcFix(r)
-				if filtered {
-					continue
-				}
+		const languageRust = 28
+
+		language := ent.Val(dwarf.AttrLanguage)
+		if language != nil && language.(int64) == languageRust {
+			rawRanges, err := rustRanges(debugInfo, ent)
+			if err != nil {
+				return nil, nil, fmt.Errorf("failed to query Rust PC ranges: %w", err)
 			}
-			if module.Name == "" {
-				ranges = append(ranges, pcRange{r[0], r[1], unit})
-			} else {
-				ranges = append(ranges, pcRange{r[0] + module.Addr, r[1] + module.Addr, unit})
+			for _, r := range rawRanges {
+				addRange([2]uint64{r.start, r.end}, r.file)
+			}
+		} else {
+			ranges1, err := debugInfo.Ranges(ent)
+			if err != nil {
+				return nil, nil, err
+			}
+			for _, r := range ranges1 {
+				addRange(r, unitName)
 			}
 		}
 		r.SkipChildren()
 	}
+	var units []*CompileUnit
+	for _, unit := range unitMap {
+		units = append(units, unit)
+	}
 	return ranges, units, nil
 }
 
+type rustRange struct {
+	start uint64
+	end   uint64
+	file  string
+}
+
+func rustRanges(debugInfo *dwarf.Data, ent *dwarf.Entry) ([]rustRange, error) {
+	// For Rust, a single compilation unit may comprise all .rs files that belong to the crate.
+	// To properly render the coverage, we need to somehow infer the ranges that belong to
+	// those individual .rs files.
+	// For simplicity, let's create fake ranges by looking at the DWARF line information.
+	var ret []rustRange
+	lr, err := debugInfo.LineReader(ent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query line reader: %w", err)
+	}
+	var startPC uint64
+	var files []string
+	for {
+		var entry dwarf.LineEntry
+		if err = lr.Next(&entry); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, fmt.Errorf("failed to parse next line entry: %w", err)
+		}
+		if startPC == 0 || entry.Address != startPC {
+			for _, file := range files {
+				ret = append(ret, rustRange{
+					start: startPC,
+					end:   entry.Address - 1,
```

Comment:
Are you sure end is inclusive?
We have this code:
```
	for _, s := range symbols {
		for ; rangeIndex < len(ranges) && ranges[rangeIndex].end <= s.Start; rangeIndex++ {
		}
```
it means range end is exclusive.
Am I missing something?

Reply:
Yes, it indeed follows that end is exclusive.